engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobFailProcessor.java

```diff
@@ -46,70 +51,82 @@ public JobFailProcessor(
-        incidentErrorMessage = jobErrorMessage;
-      }
-
-      incidentEvent.reset();
-      incidentEvent
-          .setErrorType(ErrorType.JOB_NO_RETRIES)
-          .setErrorMessage(incidentErrorMessage)
-          .setBpmnProcessId(value.getBpmnProcessIdBuffer())
-          .setProcessDefinitionKey(value.getProcessDefinitionKey())
-          .setProcessInstanceKey(value.getProcessInstanceKey())
-          .setElementId(value.getElementIdBuffer())
-          .setElementInstanceKey(value.getElementInstanceKey())
-          .setJobKey(key)
-          .setVariableScopeKey(value.getElementInstanceKey());
+  public void processRecord(final TypedRecord<JobRecord> record) {
+    final long jobKey = record.getKey();
+    final JobState.State state = jobState.getState(jobKey);
 
-      stateWriter.appendFollowUpEvent(
-          keyGenerator.nextKey(), IncidentIntent.CREATED, incidentEvent);
-    }
+    preconditionChecker
+        .check(state, jobKey)
+        .ifRightOrLeft(
+            ok -> failJob(record),
+            violation -> {
+              rejectionWriter.appendRejection(record, violation.getLeft(), violation.getRight());
+              responseWriter.writeRejectionOnCommand(
+                  record, violation.getLeft(), violation.getRight());
+            });
   }
 
-  private void acceptCommand(
-      final TypedRecord<JobRecord> command, final CommandControl<JobRecord> commandControl) {
-    final long key = command.getKey();
-    final JobRecord failedJob = jobState.getJob(key);
-    final var retries = command.getValue().getRetries();
-    final var retryBackOff = command.getValue().getRetryBackoff();
+  private void failJob(final TypedRecord<JobRecord> record) {
+    final long jobKey = record.getKey();
+    final JobRecord failJobCommandRecord = record.getValue();
+    final var retries = failJobCommandRecord.getRetries();
+    final var retryBackOff = failJobCommandRecord.getRetryBackoff();
+
+    final JobRecord failedJob = jobState.getJob(jobKey);
     failedJob.setRetries(retries);
-    failedJob.setErrorMessage(command.getValue().getErrorMessageBuffer());
+    failedJob.setErrorMessage(failJobCommandRecord.getErrorMessageBuffer());
     failedJob.setRetryBackoff(retryBackOff);
+    failedJob.setVariables(failJobCommandRecord.getVariablesBuffer());
+
     if (retries > 0 && retryBackOff > 0) {
-      final long receivedTime = command.getTimestamp();
+      final long receivedTime = record.getTimestamp();
       failedJob.setRecurringTime(receivedTime + retryBackOff);
       sideEffectWriter.appendSideEffect(
           () -> {
             jobBackoffChecker.scheduleBackOff(retryBackOff + receivedTime);
             return true;
           });
     }
-    commandControl.accept(JobIntent.FAILED, failedJob);
+    stateWriter.appendFollowUpEvent(jobKey, JobIntent.FAILED, failedJob);
+    responseWriter.writeEventOnCommand(jobKey, JobIntent.FAILED, failedJob, record);
     jobMetrics.jobFailed(failedJob.getType());
+
```

Comment:
❓ Did we omit `setFailedVariables(failedJob);` intentionally here?

Reply:
@berkaycanbc yes, this is intentional as we don't support setting variables on job failure in `8.1`.